.github/workflows/workflow.yml

```diff
@@ -0,0 +1,81 @@
+name: CI/CD
+on:
+  workflow_dispatch:
+  push:
+    branches:
+      - main
+  pull_request:
+    branches:
+      - main
+jobs:
+  Hadolint:
+    runs-on: ubuntu-22.04
+    permissions:
+      contents: read
+    steps:
+      - name: checkout
+        uses: actions/checkout@v3
+      - name: lint
+        uses: hadolint/hadolint-action@v3.1.0
+        with:
+          failure-threshold: style
+          ignore: DL3018,DL4006
+  Docker:
+    needs: hadolint
+    runs-on: ubuntu-22.04
+    permissions:
+      contents: read
+      packages: write
+    steps:
+      - name: Checkout
+        uses: actions/checkout@v3
+      - name: Setup QEMU
+        uses: docker/setup-qemu-action@v2
+      - name: Setup Docker
+        uses: docker/setup-buildx-action@v2
+      - name: Build
+        uses: docker/build-push-action@v4
+        with:
+          context: .
+          platforms: amd64
+          push: false
+          cache-to: type=local,dest=/tmp/.buildx-cache
+          cache-from: type=registry,ref=ghcr.io/utopia-planitia/kubernetes-dns-node-cache-image:cache,ignore-error=true
+      - name: Extract versions
+        if: github.event_name != 'pull_request'
+        run: |
+          set -eux
+          IPTABLES_VERSION="$(grep -Eom1 'IPTABLES_VERSION=[0-9]+\.[0-9]+\.[0-9]+\S*' Dockerfile | cut -d= -f2)"
+          echo "IPTABLES_VERSION=${IPTABLES_VERSION:?}" | tee -a "${GITHUB_ENV:?}"
+          KUBERNETES_DNS_VERSION="$(grep -Eom1 'KUBERNETES_DNS_VERSION=[0-9]+\.[0-9]+\.[0-9]+\S*' Dockerfile | cut -d= -f2)"
+          echo "KUBERNETES_DNS_VERSION=${KUBERNETES_DNS_VERSION:?}" | tee -a "${GITHUB_ENV:?}"
+      - name: Prepare metadata
+        if: github.event_name != 'pull_request'
+        id: meta
+        uses: docker/metadata-action@v4
+        with:
+          images: ghcr.io/utopia-planitia/kubernetes-dns-node-cache-image
+          tags: |
+            type=match,pattern=^(\d+\.\d+\.\d+-iptables\d+\.\d+\.\d+\S*)$,group=1,value=${{ env.KUBERNETES_DNS_VERSION }}-iptables${{ env.IPTABLES_VERSION }}
```

Comment:
What happens when we just update the Dockerfile without changing the KUBERNETES_DNS_VERSION nor the IPTABLES_VERSION?

I think we also need to introduce our own version so that the tag would be for example `1.3.0-k8sdns1.21.1-iptables1.8.8

Reply:
Yes, currently we would push to existing tags. I thought about finding a way of avoiding this but I did not come up with a solution that is simple enough. Rebuilding the image might not change the final image (except for the LABELs that include the commit hash ... :thinking:) because the artifacts of the build stages should be unchanged. We can pin the image digest where we use the image to avoid issues caused by re-used tags. And except for the full tag that includes the build version of the iptables package (eg. "-r0") reusing the tag was my intention.

I am not sure if we should introduce our own version but for simplicity reasons I would not use semver but a simple counter (similar to our cluster releases).

Reply:
Maybe we should only push when the Dockerfile has changed. But then it might be difficult when we intentionally want to re-build and push an image for some reason. And tags might still get reused when the base images get updated for example.